lib/Ceilometer/Types.hs

```diff
@@ -1 +1,111 @@
+
 module Ceilometer.Types where
+
+--import           Control.Isomorphism.Partial.Iso
+import           Data.Bimap                      (Bimap)
+import qualified Data.Bimap                      as B
+import           Data.HashMap.Strict             (HashMap)
+import qualified Data.HashMap.Strict             as H
+import           Data.Text                       (Text)
+import qualified Data.Text                       as T
+import           Data.Word
+
+import           Marquise.Client
+import           Vaultaire.Types
+
+import           Ceilometer.Types.Base
+import           Ceilometer.Types.IP
+import           Ceilometer.Types.Volume
+
+import Control.Lens.Prism
+
+type Env = (FlavorMap, SourceDict)
+
+lookupMetricName = lookupSource "metric_name"
+lookupEvent      = lookupSource "_event"
+
+type CSimple = SimplePoint
+
+-- | Views the raw @SimplePoint@ as-is.
+cSimple :: Env -> Prism' SimplePoint CSimple
+cSimple (_, sd) = prism' id parse
+  where parse raw = case (lookupMetricName sd, isEvent sd) of
+          (Just "disk.write.bytes", True) -> Just raw
+          (Just "disk.read.bytes",  True) -> Just raw
+          _                               -> Nothing
+
+data CVolume = CVolume Address TimeStamp VolumeStatus VolumeVerb Endpoint Word64
+
+-- | Views the raw @SimplePoint@ as a Ceilometer point @CVolume@.
+cVolume :: Env -> Prism' SimplePoint CVolume
+cVolume (flavours, sd) = prism' toSP fromSP
+  where toSP = undefined
+        fromSP (SimplePoint a ts v) = case (lookupMetricName sd, isEvent sd) of
+          (Just "volume.size", False) -> Just $ CVolume a ts _ _ _ _
+
+data CIPEvent = CIPEvent Address TimeStamp IPStatus IPVerb Endpoint
+
+-- | Views the raw @SimplePoint@ as a partially applied `step`
+--   to use in an aggregation of @SimplePoint@.
+--   Assuming the aggreation is a fold.
+step :: Prism' SimplePoint (acc -> acc)
+step = undefined
+
+
+
+
+-- Snapshots
+data SnapshotStatus
+data SnapshotVerb
+
+-- Images
+data ImageStatus
+data ImageVerb
+
+-- Instances
+--
+newtype Flavor = Flavor String
+  deriving (Show, Eq, Ord)
+
+type FlavorMap = Bimap Flavor Word64
+
+data InstanceStatus
+data InstanceVerb
+
+data InstancePayload = InstanceFlavor Flavor
+                     | InstanceRam    Word64
+                     | InstanceDisk   Word64
+                     | InstanceVCPU   Word64
+  deriving (Show, Eq, Ord)
+
+
+data CeilometerPoint
+  = InstancePoint       Address TimeStamp InstanceStatus                       InstancePayload
+  | IPEventPoint        Address TimeStamp IPStatus       IPVerb       Endpoint
+  | VolumeEventPoint    Address TimeStamp VolumeStatus   VolumeVerb   Endpoint Word64
+  | SnapshotEventPoint  Address TimeStamp SnapshotStatus SnapshotVerb Endpoint Word64
+  | ImageEventPoint     Address TimeStamp ImageStatus    ImageVerb             Word64
+  | StandardPoint       Address TimeStamp                                      Word64
+
+--type VaultIso = Iso SimplePoint
+
+isEvent :: SourceDict -> Bool
```

Comment:
I don't like how we had to define a bunch of partial predicates like this on what is essentially the schema of our data either. Instead, it would be nice if we can somehow parse this dictionary into a schema of the types we want.